rfcs/2022-04-09-accessibility.md

```diff
@@ -0,0 +1,495 @@
+
+### Areas to cover
+
+- [Keyboard tab order](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation-order)
+- [Using tabindex](https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute)
+- [Skip navigation](https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute)
+- [Composite widgets](https://www.w3.org/TR/wai-aria-1.2/#composite)
+  - [Roving tabindex pattern](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex)
+  - [aria-activedescendant pattern (skipped in implementation)](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_focus_activedescendant)
+- [Manual testing](#manual-testing)
+
+## Keyboard tab order
+
+> [Specification](https://www.w3.org/WAI/WCAG21/quickref/#focus-order)
+
+Tab order is important for proper navigation through a keyboard interface. 
+The tab order of the page must be logical and follow the visual order of elements on the page.
+
+**Best practices for logical tab order:**
+
+- Structure html so that the reading/navigation [order is correct](https://www.w3.org/WAI/WCAG21/Techniques/css/C27).
+- If necessary, use CSS to control the visual presentation of the elements on the page.
+- [Avoid using tabindex](https://www.w3.org/WAI/WCAG21/Techniques/failures/F44) values of 1 or greater to change the default keyboard navigation order.
+
+## Using tabindex
+
+`tabindex` attribute authors can make the element focusable and appear in [sequential focus order](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation)
+
+Setting tabindex to any non-negative integer makes elements focusable, allow or prevent them from being sequentially focusable, and determine their relative ordering
+
+When set to `0`, the element becomes focusable by keyboard and via programmatic means with `focus()`. 
+When set to `-1`, the element becomes focusable programmatically, but it does not become part of the keyboard focus order.
+
+The following table describes tabindex behavior in modern browsers:
+
+| tabindex attribute |	Focusable with mouse or programmatically |	Tab navigable |
+| :-- | :-- | :-- |
+| not present	| The user agent will decide | The user agent will decide |
+| Negative (tabindex="-1")	| Yes | No, can only be focused programmatically |
+| Zero (tabindex="0") | Yes | Yes, In tab order relative to element's position in document | 
+| Positive (tabindex="2") | Yes | Yes, tabindex value determines where this element is positioned in the tab order | 
+
+> Warning: Avoid using positive values for tabindex. Using positive values means authors will have to set (and maintain) tabindex values for all focusable elements on the page whenever they use one or more positive values for tabindex.  
```

Comment:
We discourage using positive values because positive values can cause wrong tab order 

for example: 

```
<button tabIndex="1">one</button>
<button tabIndex="2">one</button>
<button tabIndex="3">one</button>
```

The above code might look fine but since tabindex is tracked globally these buttons will always be the first ones to get focus even if they are at the end of the page, plus someone might mess up the order and cause confusion. 

>  then we shall also discourage negative values right since that will stop tab navigation

No that depends, in certain situations we want to prevent tab navigation to reduce tab stops in the page, see composite widgets section. https://github.com/razorpay/blade/blob/anu/a11y-rfc/rfcs/2022-04-09-accessibility.md#composite-widgets

We want to reduce the number of tab stops in certain elements like select-menu/combobox/radio groups etc, And let users use Arrowkeys to move between the group. 

> I'm thinking we shall discourage using tabindex altogether

In app level consumers probably wouldn’t need to use tabindex, but in component level we need to manage that to ensure focus management is handled correctly.